Marlin/src/feature/spindle_laser.cpp

```diff
@@ -96,57 +98,131 @@ void SpindleLaser::init() {
-    _set_ocr(0);
+    ocr_set(0);
   }
-#endif
 
-//
-// Set cutter ON/OFF state (and PWM) to the given cutter power value
-//
-void SpindleLaser::apply_power(const uint8_t opwr) {
-  static uint8_t last_power_applied = 0;
-  if (opwr == last_power_applied) return;
-  last_power_applied = opwr;
-  power = opwr;
+#endif  // SPINDLE_LASER_PWM
+
+// Set state for pin laser/spindle
+// :param - enable True - on; False - off
+void SpindleLaser::pin_set(const bool enable) {
+  WRITE(SPINDLE_LASER_ENA_PIN, enable ? SPINDLE_LASER_ACTIVE_STATE : !SPINDLE_LASER_ACTIVE_STATE);
+}
+
+/**
+ * Detection event
+ * 
+ * @param opwr Power value
+ */
+SPINDLE_LASER_EVENT SpindleLaser::get_event(const uint8_t opwr) {
+  SPINDLE_LASER_EVENT event;
+
+  // Conditions
   #if ENABLED(SPINDLE_LASER_PWM)
     if (cutter.unitPower == 0 && CUTTER_UNIT_IS(RPM)) {
-      ocr_off();
-      isReady = false;
-    }
-    else if (ENABLED(CUTTER_POWER_RELATIVE) || enabled()) {
-      set_ocr(power);
-      isReady = true;
-    }
-    else {
-      ocr_off();
-      isReady = false;
+      if (power == 0) {
+        return SPINDLE_LASER_EVENT::OFF;
+      } else {
+        return SPINDLE_LASER_EVENT::TO_OFF;
+      }
     }
-  #elif ENABLED(SPINDLE_SERVO)
-    MOVE_SERVO(SPINDLE_SERVO_NR, power);
-  #else
-    WRITE(SPINDLE_LASER_ENA_PIN, enabled() ? SPINDLE_LASER_ACTIVE_STATE : !SPINDLE_LASER_ACTIVE_STATE);
-    isReady = true;
   #endif
+
+  if ((power + opwr) == 0) {
+    event = SPINDLE_LASER_EVENT::OFF;
+  } else if ((power > 0) && (opwr > 0)){
+    event = SPINDLE_LASER_EVENT::ON;
+  } else if (opwr == 0) {
+    event = SPINDLE_LASER_EVENT::TO_OFF;
+  } else {
+    event = SPINDLE_LASER_EVENT::TO_ON;
+  }
+
+  return event;
+}
+
+/**
+ * Apply power for laser/spindle
+ * 
+ * Apply cutter power value for PWM, Servo, and on/off pin.
+ * 
+ * @param opwr Power value. Range 0 to MAX. When 0 disable laser/spindel.
+ */
+void SpindleLaser::apply_power(const uint8_t opwr) {
+  static uint8_t last_power_applied = 0;
+
+  switch (get_event(opwr)) {
+    case SPINDLE_LASER_EVENT::ON:
+      if (opwr == last_power_applied) break;
+      last_power_applied = opwr;
+      power = opwr;
+
+      #if ENABLED(SPINDLE_LASER_PWM)
+        ocr_set(power);
+        isReady = true;
+      #endif
+      #if ENABLED(SPINDLE_SERVO)
+        MOVE_SERVO(SPINDLE_SERVO_NR, power);
+      #endif
+      #if DISABLED(SPINDLE_SERVO)
+        pin_set(true);
+        isReady = true;
+      #endif
+
+      break;
+    case SPINDLE_LASER_EVENT::TO_ON:
+      last_power_applied = opwr;
+      power = opwr;
+
+      #if ENABLED(SPINDLE_LASER_PWM)
+        ocr_set(power);
+        isReady = true;
+      #endif
+      #if ENABLED(SPINDLE_SERVO)
+        MOVE_SERVO(SPINDLE_SERVO_NR, power);
+      #endif
+      #if DISABLED(SPINDLE_SERVO)
+        pin_set(true);
+        isReady = true;
+      #endif
+
+      power_delay(true);
+      break;
+    case SPINDLE_LASER_EVENT::OFF:
+      break;
```

Comment:
@shizacat — Is this `break` intentional?

Reply:
Yes, may be will future this state will be need. But now isn't not. May be comment it out, but it doesn't delete - for clarity. Something like this )